dubbo-config/dubbo-config-spring/src/main/java/com/alibaba/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java

```diff
@@ -18,488 +18,233 @@
-            return fieldElements;
+        if (applicationContext.containsBean(referencedBeanName)) { // Is local @Service Bean or not ?
+            // ReferenceBeanInvocationHandler's initialization has to wait for current local @Service Bean has been exported.
+            localReferenceBeanInvocationHandlerCache.put(referencedBeanName, handler);
+        } else {
+            // Remote Reference Bean should initialize immediately
+            handler.init();
         }
 
-        public Collection<ReferenceMethodElement> getMethodElements() {
-            return methodElements;
-        }
+        return handler;
     }
 
-    /**
-     * {@link Reference} {@link Method} {@link InjectionMetadata.InjectedElement}
-     */
-    private class ReferenceMethodElement extends InjectionMetadata.InjectedElement {
+    private static class ReferenceBeanInvocationHandler implements InvocationHandler {
 
-        private final Method method;
+        private final ReferenceBean referenceBean;
 
-        private final Reference reference;
+        private Object bean;
 
-        private volatile ReferenceBean<?> referenceBean;
-
-        protected ReferenceMethodElement(Method method, PropertyDescriptor pd, Reference reference) {
-            super(method, pd);
-            this.method = method;
-            this.reference = reference;
+        private ReferenceBeanInvocationHandler(ReferenceBean referenceBean) {
+            this.referenceBean = referenceBean;
         }
 
         @Override
-        protected void inject(Object bean, String beanName, PropertyValues pvs) throws Throwable {
-
-            Class<?> referenceClass = pd.getPropertyType();
-
-            referenceBean = buildReferenceBean(reference, referenceClass);
-
-            ReflectionUtils.makeAccessible(method);
-
-            method.invoke(bean, referenceBean.getObject());
-
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            return method.invoke(bean, args);
         }
 
-    }
-
-    /**
-     * {@link Reference} {@link Field} {@link InjectionMetadata.InjectedElement}
-     */
-    private class ReferenceFieldElement extends InjectionMetadata.InjectedElement {
-
-        private final Field field;
-
-        private final Reference reference;
-
-        private volatile ReferenceBean<?> referenceBean;
-
-        protected ReferenceFieldElement(Field field, Reference reference) {
-            super(field, null);
-            this.field = field;
-            this.reference = reference;
+        private void init() {
+            this.bean = referenceBean.get();
         }
+    }
 
-        @Override
-        protected void inject(Object bean, String beanName, PropertyValues pvs) throws Throwable {
-
-            Class<?> referenceClass = field.getType();
+    @Override
+    protected String buildInjectedObjectCacheKey(Reference reference, Object bean, String beanName,
```

Comment:
Multiple services referencing the same field here will cause a bug， see:  https://github.com/apache/incubator-dubbo/commit/d8282fe1b8075a453c7b44b4ae7982a4285df7b7